### ISSUE TITLE
Remove pickle due to security risk

### ISSUE THREAD

**Issue by tim-schilling** [curvedinf/hypercells#1]
Hi Curved! I like the idea behind this project. I was looking behind the curtains to see how things were being managed and I noticed that it uses Pickle. Is it possible for this dependency to be removed as can lead to execution of arbitrary code during unpickling (see the warning at https://docs.python.org/3/library/pickle.html)?

**Comment by curvedinf** [comment]
Hey Tim, that advisory is for untrusted data being unpickled. In this case, the data can be considered trusted since it is generated without user input.

In other words, the create function generates the pickled data off a QuerySet defined by the programmer, which is stored in the database, and subsequent unpicklings are impossible to alter. The only exception is if the database were compromised one could put a malicious payload in the unpickling data, but then I don't see this code execution as the largest of your issues.

Maybe I don't understand your concern? Could you elaborate if you still see an issue after my explanation?

**Comment by tim-schilling** [comment]
I think you understand the concern well. Maybe it's a difference of opinion regarding risk analysis which is fine. I appreciate you taking the time to provide a well thought out answer!

**Comment by curvedinf** [comment]
Did you have any replacements in mind when you considered this issue? I'm not opposed to changing it considering the LOE is most likely small.

**Comment by tim-schilling** [comment]
Nothing that could serve as a drop-in replacement. The more I think about it, the more I realize it's a pain to do. The utility would need to be able to regenerate the queryset. One option here is to make the interface a CBV MixIn that uses `get_queryset()` rather than loading up the data from the database.

There was some package out there that allowed serializing the `Q` objects, but I can't find it right now.

**Comment by curvedinf** [comment]
I would rather not majorly change the serialization/deserialization
methodology too much because I have done it other ways and this is the most
eloquent by far.

One thing that could be done as an optional feature to mitigate the man in
the middle attack vector is an encrypted hash also stored with the
serialization data. It can be symmetric encryption since the sender and
receiver are the same.

Basically, a new field 'data_hash' on the context. Generate the hash by
running SHA on the data, then AES with a key from settings. Upon a get for
a page, first verify the data by decrypting the hash, then run SHA on the
data and comparing the two. If it fails, raise an exception. If it passes,
continue with the request.

**Comment by tim-schilling** [comment]
You may not even need the second field. You could sign the bytes itself with [`hmac`](https://docs.python.org/3/library/hmac.html#module-hmac), then confirm the signature's validity before unpickling it.

**Comment by curvedinf** [comment]
I added a quick commit that does exactly this. I used Django's built-in SECRET_KEY since that will have the level of security that would be acceptable for most applications and requires no additional config.

Here's the commit if you would like to check it out: 
https://github.com/curvedinf/hypercells/commit/02844842ec1935c1a237f7e932dcfe430c21d9bb

**Comment by tim-schilling** [comment]
Good stuff. I like that!
